src/main/kotlin/pt/up/fe/ni/website/backend/service/RoleService.kt

```diff
@@ -1,46 +1,113 @@
-        foundActivity.role = role
-        foundActivity.activity = activity
+        foundPerActivityRole.role = role
+        foundPerActivityRole.activity = activity
 
-        foundActivity.role = role
-        foundActivity.permissions.add(permission)
-        perActivityRoleRepository.save(foundActivity)
+        foundPerActivityRole.permissions.addAll(permissions)
+        perActivityRoleRepository.save(foundPerActivityRole)
+        if (activity.associatedRoles.find { it.activity == activity } == null) {
+            role.associatedActivities.add(foundPerActivityRole)
+        }
+        // activityRepository.save(activity)
+        roleRepository.save(role)
     }
 
-    fun revokePermissionFromRoleOnActivity(role: Role, activity: Activity, permission: Permission) {
+    fun revokePermissionFromRoleOnActivity(roleId: Long, activityId: Long, permissions: Permissions) {
+        val activity = activityService.getActivityById(activityId)
+        val role = getRole(roleId)
         val foundActivity = activity.associatedRoles
             .find { it.role == role } ?: return
-
-        foundActivity.permissions.remove(permission)
+        foundActivity.permissions.removeAll(permissions)
         perActivityRoleRepository.save(foundActivity)
+        // activityRepository.save(activity)
+    }
+
+    fun createNewRole(dto: RoleDto): Role {
+        if (roleRepository.findByName(dto.name) != null) {
+            throw IllegalArgumentException(ErrorMessages.roleAlreadyExists)
+        }
+        val role = dto.create()
+
+        val latestGeneration = generationRepository.findFirstByOrderBySchoolYearDesc()
+            ?: throw IllegalArgumentException(ErrorMessages.noGenerations)
+
+        // need to set it from both sides due to testing transaction
+        role.generation = latestGeneration
+        roleRepository.save(role)
+        latestGeneration.roles.add(role)
+        generationRepository.save(latestGeneration)
+        return role
+    }
+
+    fun deleteRole(roleId: Long) {
+        val role = getRole(roleId)
+        role.generation.roles.remove(role)
+        generationRepository.save(role.generation)
+        roleRepository.delete(role)
+    }
+
+    fun addUserToRole(roleId: Long, userId: Long) {
+        val role = getRole(roleId)
+        val account = accountService.getAccountById(userId)
+        role.accounts.find { it.id == account.id }.let {
+            if (it != null) throw NoSuchElementException(ErrorMessages.userAlreadyHasRole(roleId, userId))
+        }
+        role.accounts.add(account)
+        roleRepository.save(role)
+    }
+
+    fun removeUserFromRole(roleId: Long, userId: Long) {
+        val role = getRole(roleId)
+        val account = accountService.getAccountById(userId)
+        role.accounts.find { it.id == account.id }.let {
+            if (it == null) throw NoSuchElementException(ErrorMessages.userNotInRole(roleId, userId))
+        }
```

Comment:
```kotlin
if (role.accounts.none { it.id == account.id }) {
    throw NoSuchElementException(ErrorMessages.userNotInRole(roleId, userId))
}
```
I think this makes the code a bit more readable!

Reply:
In order to make the this api call idempotent according to @BrunoRosendo suggestion I'll have to change this to:
`if(role.accounts.any { it.id == account.id }) return`